support-workers/src/main/scala/com/gu/support/workers/lambdas/CreateZuoraSubscription.scala

```diff
@@ -323,23 +323,42 @@ object DigitalSubscriptionGiftRedemption {
     val codeValidator = new GiftCodeValidator(zuoraService)
     for {
       codeValidation <- codeValidator.getStatus(redemptionData.redemptionCode, Some(state.requestId.toString))
-      subIdUpdateAction <- codeValidation match {
-        case ValidGiftCode(subscriptionId) => Future.successful((
+      termDates <- codeValidation match {
+        case ValidGiftCode(subscriptionId) => doZuoraUpdates(
           subscriptionId,
-          zuoraService.updateSubscriptionRedemptionData(subscriptionId, state.requestId.toString, state.user.id, LocalDate.now(), _)
-        ))
-        case CodeRedeemedInThisRequest(subscriptionId) => Future.successful(subscriptionId, (_: Int) => Future.successful(UpdateRedemptionDataResponse(true)))
+          zuoraService.updateSubscriptionRedemptionData(subscriptionId, state.requestId.toString, state.user.id, LocalDate.now(), _),
+          zuoraService.setupRevenueRecognition,
+          zuoraService,
+          catalogService
+        )
+        case CodeRedeemedInThisRequest(subscriptionId) => doZuoraUpdates(
+          subscriptionId,
+          (_: Int) => Future.successful(ZuoraSuccessOrFailureResponse(success = true, None)),
+          (_: Subscription, _:TermDates) => Future.successful(()),
+          zuoraService,
+          catalogService
+        )
         case otherState: CodeStatus => Future.failed(new RuntimeException(otherState.clientCode))
       }
-      (subscriptionId, updateIfNecessary) = subIdUpdateAction
-      fullGiftSubscription <- zuoraService.getSubscriptionById(subscriptionId)
-      calculatedDates <- Future.fromTry(calculateNewTermLength(fullGiftSubscription, catalogService))
-      (dates, newTermLength) = calculatedDates
-      updateDataResponse <- updateIfNecessary(newTermLength)
-      handlerResult <- Future.fromTry(buildHandlerResult(updateDataResponse, state, requestInfo, dates))
+      handlerResult <- Future.fromTry(buildHandlerResult(state, requestInfo, termDates))
     } yield handlerResult
   }
 
+  private def doZuoraUpdates(
+    subscriptionId: String,
+    updateGiftIdentityIdCall: Int => Future[ZuoraSuccessOrFailureResponse],
+    setupRevenueRecognitionCall: (Subscription, TermDates) => Future[Unit],
```

Comment:
I know I started it, but we're proliferating anonymous functions a bit here, consider a type alias or trait to gain nice naming or type checking respectively.

Reply:
We're only passing one function in now so I'm going to leave as is